parser.d.ts

```diff
@@ -0,0 +1,159 @@
+type _If<Test, Then, Else> = Test extends true ? Then : Else;
+
+export type Features = {
+  lookbehind?: boolean;
+  namedGroups?: boolean;
+  unicodePropertyEscape?: boolean;
+  unicodeSet?: boolean;
+};
+
+export type AstNodeType =
+  | "alternative"
+  | "anchor"
+  | "characterClass"
+  | "characterClassEscape"
+  | "characterClassRange"
+  | "disjunction"
+  | "dot"
+  | "group"
+  | "quantifier"
+  | "reference"
+  | "unicodePropertyEscape"
+  | "value";
+
+export type Base<T extends AstNodeType> = {
+  range: [number, number];
+  raw: string;
+  type: T;
+};
+
+export type AstNode<F extends Features = {}> =
+  | Alternative<F>
+  | Anchor
+  | CharacterClass<F>
+  | CharacterClassEscape
+  | CharacterClassRange
+  | Disjunction<F>
+  | Dot
+  | Group<F>
+  | Quantifier<F>
+  | Reference<F>
+  | UnicodePropertyEscape
+  | Value;
+
+export type RootNode<F extends Features = {}> =
+  | Exclude<AstNode<F>, CharacterClassRange | UnicodePropertyEscape>
+  | _If<F["unicodePropertyEscape"], UnicodePropertyEscape, never>;
+
+export type Anchor = Base<"anchor"> & {
+  kind: "boundary" | "end" | "not-boundary" | "start";
+};
+
+export type CharacterClassEscape = Base<"characterClassEscape"> & {
+  value: string;
+};
+
+export type Value = Base<"value"> & {
+  codePoint: number;
+  kind:
+    | "controlLetter"
+    | "hexadecimalEscape"
+    | "identifier"
+    | "null"
+    | "octal"
+    | "singleEscape"
+    | "symbol"
+    | "unicodeCodePointEscape"
+    | "unicodeEscape";
+};
+
+export type Identifier = Base<"value"> & {
+  value: string;
+};
+
+export type Alternative<F extends Features = {}> = Base<"alternative"> & {
+  body: RootNode<F>[];
+};
+
+export type CharacterClassRange = Base<"characterClassRange"> & {
+  max: Value;
+  min: Value;
+};
+
+export type UnicodePropertyEscape = Base<"unicodePropertyEscape"> & {
+  negative: boolean;
+  value: string;
+};
+
+export type CharacterClassBody =
+  | CharacterClassEscape
+  | CharacterClassRange
+  | UnicodePropertyEscape
+  | Value;
+
+export type CharacterClass<F extends Features = {}> = Base<"characterClass"> & {
+  body: CharacterClassBody[];
+  negative: boolean;
+  kind: "union" | _If<F["unicodeSet"], "intersection" | "subtraction", never>;
+};
+
+export type NonCapturingGroup<F extends Features = {}> = Base<"group"> & {
+  behavior:
+    | "ignore"
+    | "lookahead"
+    | "lookbehind"
+    | "negativeLookahead"
+    | "negativeLookbehind";
+  body: RootNode<F>[];
+};
+
+export type CapturingGroup<F extends Features = {}> = Base<"group"> & {
+  behavior: "normal";
+  body: RootNode<F>[];
+} & _If<
+    F["namedGroups"],
+    {
+      name?: Identifier;
+    },
+    {
+      name: undefined;
+    }
```

Comment:
Is this set to `undefined` or missing when the option is disabled?

Reply:
This is needed to tell TypeScript that the property definitely won't be there(/ is `undefined`) so that
```ts
if (
  nodeWithMaybeNamedGroups.type === "group" &&
  nodeWithMaybeNamedGroups.behavior === "normal"
) {
  assert<Identifier | undefined>(nodeWithMaybeNamedGroups.name);
}
```
works and more specifically
```ts
if (
  nodeWithMaybeNamedGroups.type === "group" &&
  nodeWithMaybeNamedGroups.behavior === "normal"
) {
  if (nodeWithMaybeNamedGroups.name) {
    // `name` is `Identifier` here
  }
}
```

If you move the `undefined` line above this last snippet won't work. It won't let you access `name`